### DISCUSSION TITLE
tutorials/lift-off-part4/06-resolving-a-mutation-with-errors

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#41] [course feedback]
# tutorials/lift-off-part4/06-resolving-a-mutation-with-errors

https://www.apollographql.com/tutorials/lift-off-part4/06-resolving-a-mutation-with-errors

**Comment by AkashSkySingh** [comment]
Slight mishap on the first question with the references of `err` vs `error`.

**Reply by mabuyo** [reply]
Hi there @AkashSkySingh! Are you referring to the multiple choice quiz? We'll update the quiz to be more consistent with the lesson contents. Just wanted to note that what you name the variable in the `catch` block doesn't matter, as long as you use it the same way inside the block. For example, both of the snippets below are still valid:

```js
catch (e) {
  console.log(e.extensions.response.body);
}
```

```js
catch (error) {
  console.log(error.extensions.response.body);
}
```

**Comment by LeeWeiZhang** [comment]
I write the exact same solution in code challenges, and it return me wrong.

**Reply by mabuyo** [reply]
Hey there @LeeWeiZhang! I tested the code challenge with the solution given and the tests passed. If you don't mind pasting in your solution here, I can take a look!

**Reply by LeeWeiZhang** [reply]
sure, this is my code.
```
/* your code:
// schema for reference
const typeDefs = gql`
  type Mutation {
    assignSpaceship(spaceshipId: ID!, missionId: ID!): AssignSpaceshipResponse
  }
  
  type AssignSpaceshipResponse {
    code: Int!
    success: Boolean!
    message: String!
    spaceship: Spaceship
    mission: Mission
  }
`;

// resolvers
const resolvers = {
  Mutation: {
    assignSpaceship: async (_, { spaceshipId, missionId }, { dataSources }) => {
      // start here
      try {
        const { spaceship, mission } = await dataSources.spaceAPI.assignSpaceshipToMission(spaceshipId, missionId);
        return {
          code: 200,
          success: true,
          message: `Successfully assigned spaceship ${spaceshipId} to ${missionId}`,
          spaceship: spaceship,
          mission: mission
        };
      // continue here
      } catch (error) {
        return {
          code: error.extensions.response.status,
          success: false,
          message: error.extensions.responses.body,
          spaceship: null,
          mission: null,
        }
      }
    }
  }
}
*/
```

**Reply by mabuyo** [reply]
Thanks! It looks like the culprit was this line in your solution:

```
message: error.extensions.responses.body,
```

It should be `response` instead of `responses`. However, I'm not sure why _all_ of the tests failed instead of just the relevant one (`the message property of the returned object should be set to the correct error message (dynamic, not hardcoded)`) . We'll investigate! 

In the meantime, if you change it to be `response`, then you should be able to pass the code challenge!

**Reply by LeeWeiZhang** [reply]
Ooops, silly mistake. Yes, changing `responses` to `response` works.🎉
Thanks for letting me know. 🙏

My easy guess was it throw error and couldn't have the output to run all test cases. 😁

**Comment by CodyJamesCasey** [comment]
Hey there! Is there a recommended way to generate types for the error in the catch block? Right now, it is of type 'unknown' and running my code generation command from previous parts of this series doesn't know how to resolve it. Would we have to add something for the code generation to map to a new model for the error?

**Reply by BlenderDude** [reply]
🤔 Context

The catch block error type used to be any. The issue with this is it leads to a situation where the error type is assumed by the developer when they probably shouldn't. `any` describes that the current type is _anything_ and can be used in _any way_. Something with an `any` can be passed to any function, or used in any object and Typescript will let you do it. _**This type should be avoided at all cost.**_

`unknown` describes we know _nothing_ about the current type, and need to do more analysis and "type narrowing". Unlike any,  `unknown` can't be used for anything except other functions that accept an unknown type.

😠 Why is TS mad here?

For example, in the catch block in the tutorial contains `err.extensions.response.status`, and Typescript has actually discovered a flaw in the error handling code. The `catch` block assumes `err` is of type `GraphQLError` , but we really don't have any guarantees that will be the type thrown. If something else throws, say something that throws a base `Error` instance, the call to `.extensions.response` will throw an error as those properties don't exist. And now that the `catch` block has thrown, the error is now gonna wreck your GraphQL Request.

💡 How can it be fixed? Type narrowing!

For the catch clause, we need to introduce some type narrowing to tell TS that we will know at runtime what the type is. Type narrowing can be done in many ways, but in this case we need to use the `instanceof` type checker! Let's see what that does 👇🏻 
```ts
try {
...
} catch (err) {
  if (err instanceof GraphQLError) {
    return {
      code: error.extensions.response.status,
      success: false,
      message: error.extensions.responses.body,
      spaceship: null,
      mission: null,
    }
  }
  return {
    code: 500,
    success: false,
    message: 'Internal Server Error',
    spaceship: null,
    mission: null,
  }
}
```
Now with the type guard, the `err` type is _narrowed_ from `unknown` to `GraphQLError` but only inside that block. If `err instanceof GraphQLError` resolves to true, we know for a _fact_ that that the unknown type is actually a `GraphQLError` during the runtime, so TS reassigns the type.

We still have the dangling case where we need to return, but there we can just an send Internal Server Error. We could also do a second `if` statement that checks `err instanceof Error` and return `err.message` for the GraphQL message, but if we have a generic error, we likely want to fail the request and not expose that error to the user anyway 🙂 

Hope this helps!

**Reply by MarvinXu** [reply]
This is still not correctly typed. `response` is of type `unknow`

**Reply by tanujabhatnagar** [reply]
@BlenderDude ... I tried , it says 

{
  "errors": [
    {
      "message": "GraphQLError is not defined",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "incrementTrackViews"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "ReferenceError: GraphQLError is not defined",
          "    at Object.incrementTrackViews (/Users/tanujabhatnagar/Desktop/odyssey-lift-off-part4/server/src/resolvers.js:30:28)",
          "    at field.resolve (/Users/tanujabhatnagar/Desktop/odyssey-lift-off-part4/server/node_modules/@apollo/server/dist/cjs/utils/schemaInstrumentation.js:41:28)",
          "    at executeField (/Users/tanujabhatnagar/Desktop/odyssey-lift-off-part4/server/node_modules/graphql/execution/execute.js:492:20)",
          "    at /Users/tanujabhatnagar/Desktop/odyssey-lift-off-part4/server/node_modules/graphql/execution/execute.js:377:22",
          "    at promiseReduce (/Users/tanujabhatnagar/Desktop/odyssey-lift-off-part4/server/node_modules/graphql/jsutils/promiseReduce.js:23:9)",
          "    at executeFieldsSerially (/Users/tanujabhatnagar/Desktop/odyssey-lift-off-part4/server/node_modules/graphql/execution/execute.js:373:43)",
          "    at executeOperation (/Users/tanujabhatnagar/Desktop/odyssey-lift-off-part4/server/node_modules/graphql/execution/execute.js:347:14)",
          "    at execute (/Users/tanujabhatnagar/Desktop/odyssey-lift-off-part4/server/node_modules/graphql/execution/execute.js:136:20)",
          "    at executeIncrementally (/Users/tanujabhatnagar/Desktop/odyssey-lift-off-part4/server/node_modules/@apollo/server/dist/cjs/incrementalDeliveryPolyfill.js:47:34)",
          "    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"
        ]
      }
    }
  ]
}

**Reply by mabuyo** [reply]
@tanujabhatnagar - that error usually means the import is missing. Make sure you add it to the top of the file:

```ts
import { GraphQLError } from 'graphql';
```

**Comment by megha8473** [comment]
hi, i am getting this error:


  "data": {
    "incrementTrackViews": {
      "code": 500,
      "success": false,
      "message": "response is not defined",
      "track": null
    }
  }

code:


   Mutation:{
      // increments a track's numberOfViews property
 incrementTrackViews: async (_,{id},{dataSources})=>{
     try {
          const track = await dataSources.trackAPI.incrementTrackViews(id);
           if (!response) throw new Error("No response received from API");
          return {
          code: 200,
          success: true,
          message: `Successfully incremented number of views for track ${id}`,
          track
      };
  } catch (err) {
      console.error("Error in incrementTrackViews:", err); // Debugging info
        let status = 500; // Default status
let message = "An unknown error occurred"; // Default message

// Check if 'extensions' exists and has a 'response' object
if (err.extensions && err.extensions.response) {
  status = err.extensions.response.status || 500;
  message = err.extensions.response.body || "No response body provided";
} else {
  message = err.message || "Unknown error occurred"; // Fallback to default message
}
  return {
    code: status,
    success: false,
    message: message,
    track: null
  };
} 
},
},

**Reply by mabuyo** [reply]
Hey there @megha8473! The error message says `response is not defined` which should help you out! In your code:

```
const track = await dataSources.trackAPI.incrementTrackViews(id);
if (!response) throw new Error("No response received from API");
```
You have not defined a `response` variable. Did you mean to say `!track`?
